### DISCUSSION TITLE
Can I customize PenaltyManager and CostEvaluator? If so, how?

### DISCUSSION THREAD

**Discussion by kobeIsMyYouth** [PyVRP/PyVRP#583] [Q&A] [answered]
Hello All!

I'm super excited to see this excellent work. I am currently doing research on constraint handling techniques in solving constrained optimization problems using metaheuristic algorithms. I would like to use PyVRP to test different constraint handling techniques, which involves reimplementation of penaltyManger and costEvaluation. I would like to ask if I can reimplement penaltyManger and costEvaluation. If possible, how can I do it?

Thank you very much for your time.

Best regards,
Ting

**Comment by N-Wouda** [comment, marked as answer]
Hi @kobeIsMyYouth,

The `PenaltyManager` is defined in `pyvrp/PenaltyManager.py`, and the `CostEvaluator` in `pyvrp/cpp/CostEvaluator.h` (and the corresponding source file). You can modify those in a fork and then install this forked version of PyVRP in your project. An example of how to install a modified version of PyVRP from git using poetry can be found [here](https://github.com/leonlan/dynamic-dispatch-waves/blob/master/pyproject.toml#L25). Hope that helps.

**Reply by bmorlo** [reply]
Hi @N-Woud,

I thought I'd add my questions here.

I was able to fork clone and run my forked and cloned pyvrp repository with poetry. What I am struggling with is how I can now inspect, edit, compile and import this package (see my screenshot with my own .toml file on the right). Is there some kind of manual regarding how this is done? I switched to Linux now and I am working with VSCode. My best guess was to just read through poetry's website. Should I just add the local path to the PyVRP folder to my .toml under poetry dependencies? Maybe you have a tip or two.

Thanks & cheers as always.

![ksnip_20240523-173711](https://github.com/PyVRP/PyVRP/assets/164156189/f49b8d45-4b2d-4ec5-8cad-d27c3721cae3)

**Reply by N-Wouda** [reply]
Presumably yes, if I understand this correctly: https://python-poetry.org/docs/dependency-specification/#path-dependencies.
